cue/validate.go

```diff
@@ -0,0 +1,101 @@
+package cue
+
+import (
+	_ "embed"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"os"
+
+	"cuelang.org/go/cue"
+	"cuelang.org/go/cue/cuecontext"
+	cueerror "cuelang.org/go/cue/errors"
+	"cuelang.org/go/encoding/yaml"
+)
+
+var (
+	//go:embed flipt.cue
+	cueFile []byte
+)
+
+// ValidateBytes takes a slice of bytes, and validates them against a cue definition.
+func ValidateBytes(b []byte) error {
+	cctx := cuecontext.New()
+
+	return validateHelper(b, cctx)
+}
+
+func validateHelper(b []byte, cctx *cue.Context) error {
+	v := cctx.CompileBytes(cueFile)
+
+	f, err := yaml.Extract("", b)
+	if err != nil {
+		return err
+	}
+
+	yv := cctx.BuildFile(f, cue.Scope(v))
+	yv = v.Unify(yv)
+
+	return yv.Validate()
+}
+
+// CueError is a collection of fields that represent positions in files where the user
+// has made some kind of error.
+type CueError struct {
+	Filename string `json:"filename,omitempty"`
+	Line     int    `json:"line"`
+	Column   int    `json:"column"`
+	Error    string `json:"error"`
+}
+
+func getCueErrors(err error, filename string) []CueError {
+	cerrs := make([]CueError, 0)
+
+	ce := cueerror.Errors(err)
+
+	for _, m := range ce {
+		fp := m.InputPositions()[0]
```

Comment:
It seems like the first `InputPosition` is the one that gives the correct line and column number from the yaml when it is invalid. I am not sure why tho...

The other Input positions in this slice give you the line and column number from the cue file that produced the invalidation.

Reply:
Nice, I think that makes sense (kind of).
That is has validation for all the sources.

Reply:
This might need a length check just for robustness